### PR TITLE
feat: web streams based encoding/csv

### PR DESCRIPTION
**Pull request by uki00a**
This PR tries to implement web streams based `encoding/csv` module.

I'm not very familiar with web streams, so any feedback would be appreciated!

Part of #1986

## Changes

- Added `encoding/csv/stream.ts`
- Added `encoding/csv/_io.ts` to remove duplicate codes between `encoding/csv.ts` and `encoding/csv/stream.ts`

### CODE REVIEW

File: encoding/csv/stream.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    },
+  });
+}
+
+export interface CSVStreamOptions {
+  separator?: string;
+  comment?: string;
+}
+
+export class CSVStream {
```
Comment (crowlKats) [inline encoding/csv/stream.ts:34]: You could rather make `CSVStream` extend `TransformStream`, then the only thing you have to do is `super()` with the transformstream options; and the variables you defined can just be private properties.

This isnt necessary, but imo just a lot cleaner & direct

File: encoding/csv/stream.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
+import { Buffer, BufReader } from "../../io/buffer.ts";
```
Comment (crowlKats) [inline encoding/csv/stream.ts:1]: Now, this is going to be the problem: any of the new streams based APIs are not allowed to depend on any Deno.Reader/Writer APIs. You'll have to wait for #1970 to land

**Review by crowlKats — CHANGES_REQUESTED** [review]

File: encoding/csv/stream.ts (hunk 2 above)
Comment (crowlKats) [reply encoding/csv/stream.ts:1]: @uki00a #1970 landed, so this should be now doable. `BufReader` & `BufWriter` dont exist anymore now (as they dont make sense for web streams). However, i see that csv also depends on textproto, which hasnt been ported yet

Comment (uki00a) [reply encoding/csv/stream.ts]: Rewrote `CSVStream` not to depend on `Deno.Reader` in [c573f87](https://github.com/denoland/deno_std/pull/1993/commits/c573f876d6bbf7e85e915b1b0c724594f0495ef9) :slightly_smiling_face:

**Review by lucacasonato — COMMENTED** [review]
I think the `CSVStream` should be a `TransformStream<string, string[]>` rather than `TransformStream<Uint8Array, string[]>`. The text decoding can be done with `TextDecoderStream` by the user.

File: encoding/csv/_io.ts [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+      this.message = `record on line ${line}: ${message}`;
+    } else if (start !== line) {
+      this.message =
+        `record on line ${start}; parse error on line ${line}, column ${column}: ${message}`;
+    } else {
+      this.message =
+        `parse error on line ${line}, column ${column}: ${message}`;
+    }
+  }
+}
```
Comment (crowlKats) [inline encoding/csv/_io.ts:236]: wouldn't it make more sense to use syntaxerror? I know it doesnt have `startLine`, but creating new error types is a bit... eh. If deemed necessary to keep, then extend syntaxerror

**Review by crowlKats — COMMENTED** [review]
except the nitpick, and luca's comment, this looks good

**Review by crowlKats — APPROVED** [review]
LGTM, thanks! the PR is still a draft; do you still have things to add? if not, i'd like to get this landed soon

**Comment by uki00a** [comment]
@crowlKats Thanks for the review! I'd like to add more tests later in a separate PR, since the tests are a little low

**Pull request merged**